### PR TITLE
Allow more types in length, isempty

### PR DESCRIPTION
**Pull request by thofma**

### CODE REVIEW

File: src/generic/Matrix.jl [hunk 1]
```diff
@@ -184,13 +184,13 @@ ncols(a::MatrixElem) = size(a.entries, 2)
     length(a::Generic.MatrixElem)
 > Return the number of entries in the given matrix.
 """
-length(a::MatrixElem) = nrows(a) * ncols(a)
+length(a::AbstractAlgebra.MatElem) = nrows(a) * ncols(a)
```
Comment (rfourquet) [inline src/generic/Matrix.jl]: Actually, `AbstractAlgebra.MatElem <: Generic.MatrixElem`. With this change, if I'm not mistaken, you make `length` applicable to less types (you basically remove `length(::MatAlgElem)`)

Comment (thofma) [reply src/generic/Matrix.jl]: No, I don't think this is true.
```julia
julia> supertype(fmpz_mat)
MatElem{fmpz}

julia> supertype(Generic.Mat{fmpz})
MatElem{fmpz}

julia> supertype(MatElem{fmpz})
ModuleElem{fmpz}
```

Comment (thofma) [reply src/generic/Matrix.jl]: Sorry, I am confused.

Comment (rfourquet) [reply src/generic/Matrix.jl]: Indeed, but `MatrixElem` is `MatrixElem = Union{MatAlgElem{T}, MatElem{T}}`, so it doesn't appear via a `supertype` call. I needed to write down a graph with these types (with `<:` relationship) to understand better how they relate.

Comment (rfourquet) [reply src/generic/Matrix.jl]: I wrote this comment before seeing your last answer, I didn't reply "indeed" to you feeling confused!

**Comment by thofma** [comment]
Sorry for the confusion.

File: src/generic/Matrix.jl (hunk 1 above)
Comment (kalmarek) [reply src/generic/Matrix.jl]: @rfourquet Could you make the graph and the script somewhere available? I'd be best to include it into documentation at some point.

Comment (rfourquet) [reply src/generic/Matrix.jl]: I did it first by hand (just the the fiew types under `MatrixElem`), but I have indeed started to work on a script, I will share when it's a bit less buggy (the Julia type system is quite complex for this task!)

Comment (rfourquet) [reply src/generic/Matrix.jl]: Alright, for this small example here is a preview: in red are types from `AbstractAlgebra.Generic`, in blue from `AbstractAlgebra`. Dashed borders means unexported, bold means abstract type.

Comment (kalmarek) [reply src/generic/Matrix.jl]: looks good! is there a reason why there is `Mat{T}`? or the graph is incomplete?

Comment (rfourquet) [reply src/generic/Matrix.jl]: > is there a reason why there is Mat{T}?

What is wrong with it?

**Pull request closed without merging**